src/txrequest.cpp

```diff
@@ -0,0 +1,729 @@
+        )) {}
+
+    // Disable copying and assigning (a default copy won't work due the stateful ByTxHashViewExtractor).
+    Impl(const Impl&) = delete;
+    Impl& operator=(const Impl&) = delete;
+
+    void DisconnectedPeer(NodeId peer)
+    {
+        auto& index = m_index.get<ByPeer>();
+        auto it = index.lower_bound(ByPeerView{peer, false, UINT256_ZERO});
+        while (it != index.end() && it->m_peer == peer) {
+            // Check what to continue with after this iteration. Note that 'it' may change position, and
+            // std::next(it) may be deleted in the process, so this needs to be decided beforehand.
+            auto it_next = (std::next(it) == index.end() || std::next(it)->m_peer != peer) ?
+                index.end() : std::next(it);
+            // If the announcement isn't already COMPLETED, first make it COMPLETED (which will mark other
+            // CANDIDATEs as CANDIDATE_BEST, or delete all of a txhash's announcements if no non-COMPLETED ones are
+            // left).
+            if (MakeCompleted(m_index.project<ByTxHash>(it))) {
+                // Then actually delete the announcement (unless it was already deleted by MakeCompleted).
+                Erase<ByPeer>(it);
+            }
+            it = it_next;
+        }
+    }
+
+    void ForgetTxHash(const uint256& txhash)
+    {
+        auto it = m_index.get<ByTxHash>().lower_bound(ByTxHashView{txhash, State::CANDIDATE_DELAYED, 0});
+        while (it != m_index.get<ByTxHash>().end() && it->m_txhash == txhash) {
+            it = Erase<ByTxHash>(it);
+        }
+    }
+
+    void ReceivedInv(NodeId peer, const GenTxid& gtxid, bool preferred,
+        std::chrono::microseconds reqtime)
+    {
+        // Bail out if we already have a CANDIDATE_BEST announcement for this (txhash, peer) combination. The case
+        // where there is a non-CANDIDATE_BEST announcement already will be caught by the uniqueness property of the
+        // ByPeer index when we try to emplace the new object below.
+        if (m_index.get<ByPeer>().count(ByPeerView{peer, true, gtxid.GetHash()})) return;
+
+        // Try creating the announcement with CANDIDATE_DELAYED state (which will fail due to the uniqueness
+        // of the ByPeer index if a non-CANDIDATE_BEST announcement already exists with the same txhash and peer).
+        // Bail out in that case.
+        auto ret = m_index.get<ByPeer>().emplace(gtxid, peer, preferred, reqtime, m_current_sequence);
+        if (!ret.second) return;
+
+        // Update accounting metadata.
+        ++m_peerinfo[peer].m_total;
+        ++m_current_sequence;
+    }
+
+    //! Find the GenTxids to request now from peer.
+    std::vector<GenTxid> GetRequestable(NodeId peer, std::chrono::microseconds now)
+    {
+        // Move time.
+        SetTimePoint(now);
+
+        // Find all CANDIDATE_BEST announcements for this peer.
+        std::vector<std::pair<SequenceNumber, const Announcement*>> selected;
```

Comment:
We can avoid using an awkward pair here by constructing a vector of pointers and then use a custom comparator function to sort:

```diff
--- a/src/txrequest.cpp
+++ b/src/txrequest.cpp
@@ -101,6 +101,9 @@ struct Announcement {
         m_is_wtxid(gtxid.IsWtxid()), m_state(uint8_t(State::CANDIDATE_DELAYED)) {}
 };
 
+bool seq_comparator(const Announcement* ann1, const Announcement* ann2) { return ann1->m_sequence < ann2->m_sequence; }
+GenTxid gtxid_extractor(const Announcement* ann) { return GenTxid{ann->m_is_wtxid, ann->m_txhash}; }
+
 //! Type alias for priorities.
 using Priority = uint64_t;
 
@@ -581,21 +584,19 @@ public:
         SetTimePoint(now);
 
         // Find all CANDIDATE_BEST announcements for this peer.
-        std::vector<std::pair<SequenceNumber, const Announcement*>> selected;
+        std::vector<const Announcement*> selected;
         auto it_peer = m_index.get<ByPeer>().lower_bound(ByPeerView{peer, true, UINT256_ZERO});
         while (it_peer != m_index.get<ByPeer>().end() && it_peer->m_peer == peer &&
             it_peer->GetState() == State::CANDIDATE_BEST) {
-            selected.emplace_back(it_peer->m_sequence, &*it_peer);
+            selected.emplace_back(&*it_peer);
             ++it_peer;
         }
 
         // Return them, sorted by sequence number.
-        std::sort(selected.begin(), selected.end());
+        std::sort(selected.begin(), selected.end(), seq_comparator);
         std::vector<GenTxid> ret;
         ret.reserve(selected.size());
-        for (const auto& item : selected) {
-            ret.emplace_back(item.second->m_is_wtxid, item.second->m_txhash);
-        }
+        std::transform(selected.begin(), selected.end(), std::back_inserter(ret), gtxid_extractor);
         return ret;
     }

```

I'm not sure if that's more or less clear code, but thought I'd throw it out there as a suggestion.

Reply:
Done, with lambdas instead of separate function (as they're only used in one place).

Reply:
Hmm, in bad cases (where GetRequestable is returning a lot of txhashes, there are a lot of announcements, and the relevant announcements are fragmented) that would be a fair bit less cache friendly than the old code (sort would be re-accessing each entry an additional `O(log(selected.size()))` times). But in reality the 5000 ann limit means at worst it just moves from sorting an 80kB array in L1 to sorting a 40kB array based on 800kB of data in L2 cache, so it probably doesn't matter. (Or rather, it's probably not even worth gathering the performance data that would be needed to make a decision based on one being faster or not)

Reply:
> Or rather, it's probably not even worth gathering the performance data that would be needed to make a decision based on one being faster or not.

Yeah.

Reply:
> Done, with lambdas instead of separate functions (as they're only used in one place).

Ha. I went the other way - did this with lambdas first and then moved them to named functions because the lines were getting a bit long. Either way seems fine.

> it's probably not even worth gathering the performance data that would be needed to make a decision based on one being faster or not

Yeah, I can see this being incrementally worse performance in the worst case, but I think readability wins.